### ISSUE TITLE
Slicing optimization - Use atAxisIndex when possible

### ISSUE THREAD

**Issue by mratsim** [mratsim/Arraymancer#232]
In many cases, we only need to slice a single dimension with a unit step, for example:

- For minibatch
  https://github.com/mratsim/Arraymancer/blob/249218eb64476c5aec34ca15d0dcf0955fbb418c/examples/ex01_xor_perceptron_from_scratch.nim#L43-L44

- For RNNs
    https://github.com/mratsim/Arraymancer/blob/c2b9748728c5986043cb327ad90689f7e41fe485/src/nn_primitives/recurrent/nnp_gru.nim#L61

- For Softmax

  https://github.com/mratsim/Arraymancer/blob/249218eb64476c5aec34ca15d0dcf0955fbb418c/src/nn_primitives/nnp_softmax.nim#L34-L40

- For convolution
  https://github.com/mratsim/Arraymancer/blob/249218eb64476c5aec34ca15d0dcf0955fbb418c/src/nn_primitives/fallback/conv.nim#L141

Benchmarking the XOR example shows that 2% of the time is spent slicing for the minibatch.
Some complex architectures or loss functions would probably require much more slicing (sequence to sequence for example).

![2018-05-11_02-03-31](https://user-images.githubusercontent.com/22738317/39900067-b0e566fa-54bf-11e8-8147-8ea82a06b22a.png)

Further investigation in Assembly-land shows that the culprits are memcpy and integer division
![image](https://user-images.githubusercontent.com/22738317/39900130-09f8e208-54c0-11e8-825e-94bf289710c5.png)

The new length parameter for atAxisIndex introduced by https://github.com/mratsim/Arraymancer/commit/f95bfc12e1c435a1d3a28bc438e659fd3f01fc67 covers this use case and would avoid the following overhead:
  - for looping on all slices with:
      - offset update
      - stride computation
      - division to determine shape

https://github.com/mratsim/Arraymancer/blob/249218eb64476c5aec34ca15d0dcf0955fbb418c/src/tensor/private/p_accessors_macros_read.nim#L24-L51

vs

https://github.com/mratsim/Arraymancer/blob/f95bfc12e1c435a1d3a28bc438e659fd3f01fc67/src/tensor/accessors.nim#L29-L37

### Solutions

  - The slicing macro should dispatch to atAxisIndex for single dim slices.
  - Internal procs should use atAxisIndex instead of slices (at the expense of readability)
